### PR TITLE
I1 patron

### PR DESCRIPTION
**Pull request by haynescrossley**
Refactored Patron Class.
Author - Haynes
Mediator - @elliotmitchum 
Reviewer - @amrmosit 

Please review Patron class changes before merging

Thanks,
Haynes

### CODE REVIEW

**Comment by elliotmitchum** [comment]
@amrmosit  This is ready for your review

**Comment by haynescrossley** [comment]
Changed to draft for now. Will update PR request once Calendar is reviewed and merged into main. Will then update Patron branch with the latest version of Master and re-request PR.

**Comment by haynescrossley** [comment]
Merged latest version of master into this branch. Includes updated Library, Item, Loan and Calendar classes. I resolved all conflicts from the merge. Hopefully there are no issues.

Ready for review @elliotmitchum

**Comment by elliotmitchum** [comment]
@amrmosit This is ready for your review and approval

File: src/library/entities/Patron.java [hunk 1]
```diff
@@ -1,118 +1,116 @@
...
+        this.lastName = lastName;
+        this.emailAddress = emailAddress;
+        this.phoneNumber = phoneNumber;
+        this.patronId = memberId;
+        this.currentLoans = new HashMap<>();
+    }
+
+    
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
```
Comment (amrmosit) [inline src/library/entities/Patron.java]: PR#36 R1-D1-CSG#2: All variable names are to be meaningful

File: src/library/entities/Patron.java [hunk 2]
```diff
@@ -1,118 +1,116 @@
...
+        sb.append("Patron:  ").append(patronId).append("\n")
+          .append("  Name:  ").append(firstName).append(" ").append(lastName).append("\n")
+          .append("  Email: ").append(emailAddress).append("\n")
+          .append("  Phone: ").append(phoneNumber)
+          .append("\n")
+          .append(String.format("  Fines Owed :  $%.2f", finesOwing))
+          .append("\n");
+        
+        for (Loan LoAn : currentLoans.values()) {
```
Comment (amrmosit) [inline src/library/entities/Patron.java]: PR#36 R1-D2-CSG#2: All variable names are to start with a lowercase letter and to be in camelBack
hint:LoAn to be loan.

File: src/library/entities/Patron.java [hunk 3]
```diff
@@ -1,118 +1,116 @@
...
+        return new ArrayList<Loan>(currentLoans.values());
+    }
+
+    
+    public int getNumberOfCurrentLoans() {
+        return currentLoans.size();
+    }
+
+    
+    public double finesOwed() {
```
Comment (amrmosit) [inline src/library/entities/Patron.java]: PR#36 R1-D3-CSG#3: All methods are to be meaningful. Method names should be verbs or verb phrases
hint: finesOwed to be getFinesOwed

File: src/library/entities/Patron.java [hunk 4]
```diff
@@ -1,118 +1,116 @@
...
+        return finesOwing;
+    }
+
+    
+    public void takeOutLoan(Loan loan) {
+        if (!currentLoans.containsKey(loan.getId())) {
```
Comment (amrmosit) [inline src/library/entities/Patron.java]: PR#36 R1-D4-CSG#10: The argument list of any metho calls MUST NOT contain any other method calls
hint: add local variable Long loanId = loan.getId();
if(!currentLoans.containsKey(loanId){ // same for line 69

File: src/library/entities/Patron.java [hunk 5]
```diff
@@ -1,118 +1,116 @@
...
+        else {
+            throw new RuntimeException("Duplicate loan added to member");
+        }
+    }
+
+
+    public void dischargeloan(Loan loan) {
+        if (currentLoans.containsKey(loan.getId())) {
+            currentLoans.remove(loan.getId());
+        }
```
Comment (amrmosit) [inline src/library/entities/Patron.java:82]: PR#36 R1-D5-CSG#10: The argument list of any metho calls MUST NOT contain any other method calls
same hint as PR#36 R1-D4-CSG#10

**Review by amrmosit — CHANGES_REQUESTED** [review]
There are 5 Defects to be resolved

File: src/library/entities/Patron.java (hunk 1 above)
Comment (haynescrossley) [reply src/library/entities/Patron.java]: [Fixed here with this commit](https://github.com/elliotmitchum/ITC205/pull/36/commits/bb18cb3ecee12555f839f8c7e2bd0f7d0e383774)

File: src/library/entities/Patron.java (hunk 2 above)
Comment (haynescrossley) [reply src/library/entities/Patron.java]: [Fixed here with this commit](https://github.com/elliotmitchum/ITC205/pull/36/commits/a3ed401a1c8b98aa6e2231bb45909088ecaa6e7e)

File: src/library/entities/Patron.java (hunk 3 above)
Comment (haynescrossley) [reply src/library/entities/Patron.java]: [Fixed with this commit](https://github.com/elliotmitchum/ITC205/pull/36/commits/b0581c296e924a588632c7da322c6d5bd237f584)

File: src/library/entities/Patron.java (hunk 4 above)
Comment (haynescrossley) [reply src/library/entities/Patron.java]: [Fixed with this commit](https://github.com/elliotmitchum/ITC205/pull/36/commits/20394d2e74ee4cdc1916fef8b02a59b24cb7e928)

File: src/library/entities/Patron.java (hunk 5 above)
Comment (haynescrossley) [reply src/library/entities/Patron.java:82]: [Fixed here with this commit](https://github.com/elliotmitchum/ITC205/pull/36/commits/3532e98039fc8d9bdc3e56579303ac073dcd90ad)

**Comment by haynescrossley** [comment]
@amrmosit I've made changes based on your feedback. Please review.

**Review by amrmosit — APPROVED** [review]
Thanks. All good. 
Branch i1-patron is reviewed and approved. 
@elliotmitchum @haynescrossley

**Comment by elliotmitchum** [comment]
Looks good, builds and runs!

**Pull request merged**